sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,621 @@
+        state->intrs = 0;
+        return EIO;
+      }
+      m = state->intrs;
+      mask &= ~state->intrs;
+      state->intrs &= ~m;
+    }
+  }
+  return 0;
+}
+
+static int bcmemmc_wait(device_t *cdev, emmc_wait_flags_t wflags) {
+  assert(cdev->parent && cdev->parent->driver == &bcmemmc_driver);
+
+  uint32_t mask = emmc_wait_flags_to_hwflags(wflags);
+  return bcmemmc_intr_wait(cdev->parent, mask);
+}
+
+static intr_filter_t bcmemmc_intr_filter(void *data) {
+  bcmemmc_state_t *state = (bcmemmc_state_t *)data;
+  resource_t *emmc = state->emmc;
+  WITH_SPIN_LOCK (&state->slock) {
+    uint32_t r = b_in(emmc, BCMEMMC_INTERRUPT);
+    state->intrs = r;
+    /* Interrupts need to be cleared manually */
+    b_out(emmc, BCMEMMC_INTERRUPT, r);
+    /* Wake up the thread if all expected interrupts have been received */
+    cv_signal(&state->cv_intr);
+  }
+  return IF_FILTERED;
+}
+
+/* This seems to be the default frequency of the clk_emmc.
+ * Preferably, it should be somewhere between 50MHz and 100MHz, but changing
+ * it requires messing around with Clock Manager, which at the moment is beyound
+ * the scope of this driver.
+ */
+#define GPIO_CLK_EMMC_DEFAULT_FREQ 41666666
+
+/* Find a divisor that provides the best approximation */
+static uint32_t bcmemmc_clk_approx_divisor(uint32_t clk, uint32_t f) {
+  int32_t c1 = clk / f;
+  if (c1 == 0)
+    c1++;
+  int32_t c2 = c1 + 1;
+  int32_t c =
+    abs((int32_t)f - (int32_t)clk / c1) < abs((int32_t)f - (int32_t)clk / c2)
+      ? c1
+      : c2;
+  return (uint32_t)c;
+}
+
+/* Clock divisor mask for BCMEMMC_CONTROL1 */
+#define BCMEMMC_CLKDIV_MASK 0xffff003f
+
+static void bcmemmc_clk_div(bcmemmc_state_t *state, uint32_t f) {
+  resource_t *emmc = state->emmc;
+
+  uint32_t clk = GPIO_CLK_EMMC_DEFAULT_FREQ;
+  uint32_t divisor = bcmemmc_clk_approx_divisor(clk, f);
+  uint32_t lo = (divisor & 0x00ff) << 8;
+  uint32_t hi = (divisor & 0x0300) >> 2;
+  b_out(emmc, BCMEMMC_CONTROL1,
+        (b_in(emmc, BCMEMMC_CONTROL1) & BCMEMMC_CLKDIV_MASK) | lo | hi);
+  klog("e.MMC: clock set to %luHz / %lu (requested %luHz)", clk, divisor, f);
+}
+
+/**
+ * set SD clock to frequency in Hz (approximately), divided mode
+ */
+static int32_t bcmemmc_clk(device_t *dev, uint32_t f) {
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *emmc = state->emmc;
+  int32_t cnt = 100000;
+
+  while ((b_in(emmc, BCMEMMC_STATUS) & (SR_CMD_INHIBIT | SR_DAT_INHIBIT)) &&
+         cnt--)
+    delay(3);
+  if (cnt <= 0) {
+    klog("e.MMC ERROR: timeout waiting for inhibit flag");
+    return ETIMEDOUT;
+  }
+
+  b_clr(emmc, BCMEMMC_CONTROL1, C1_CLK_EN);
+  /* host_version <= HOST_SPEC_V2 needs a power-of-two divisor. It would require
+   * a different calculation method. */
+  assert(state->host_version > HOST_SPEC_V2);
+  bcmemmc_clk_div(state, f);
+  b_set(emmc, BCMEMMC_CONTROL1, C1_CLK_EN);
+
+  cnt = 10000;
+  while (!(b_in(emmc, BCMEMMC_CONTROL1) & C1_CLK_STABLE) && cnt--)
```

Comment:
I wish I fully undestood the reason behind this code. This is a relic from the original driver which used polling instead of interrupts, and it miiiight be unnecessary, but I think I'll keep it until I test the driver on a physical device.

Reply:
A made a little comment.